bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/dto/MeterModelBlock.java

```diff
@@ -0,0 +1,391 @@
+
+    /**
+     * Phase B exported apparent energy
+     */
+    public Optional<Long> acExportedApparentEnergyPhaseB;
+
+    /**
+     * Phase C exported apparent energy
+     */
+    public Optional<Long> acExportedApparentEnergyPhaseC;
+
+    /**
+     * Total imported apparent energy
+     */
+    public Optional<Long> acImportedApparentEnergyTotal;
+
+    /**
+     * Phase A imported apparent energy
+     */
+    public Optional<Long> acImportedApparentEnergyPhaseA;
+
+    /**
+     * Phase B imported apparent energy
+     */
+    public Optional<Long> acImportedApparentEnergyPhaseB;
+
+    /**
+     * Phase C imported apparent energy
+     */
+    public Optional<Long> acImportedApparentEnergyPhaseC;
+
+    /**
+     * Apparent Energy Scale Factor
+     */
+    public Optional<Short> acApparentEnergySF;
+
+    /**
+     * Quadrant 1: Total imported reactive energy
+     */
+    public Optional<Long> acImportedReactiveEnergyQ1Total;
+
+    /**
+     * Quadrant 1: Phase A imported reactive energy
+     */
+    public Optional<Long> acImportedReactiveEnergyQ1PhaseA;
+
+    /**
+     * Quadrant 1: Phase B imported reactive energy
+     */
+    public Optional<Long> acImportedReactiveEnergyQ1PhaseB;
+
+    /**
+     * Quadrant 1: Phase C imported reactive energy
+     */
+    public Optional<Long> acImportedReactiveEnergyQ1PhaseC;
+
+    /**
+     * Quadrant 2: Total imported reactive energy
+     */
+    public Optional<Long> acImportedReactiveEnergyQ2Total;
+
+    /**
+     * Quadrant 2: Phase A imported reactive energy
+     */
+    public Optional<Long> acImportedReactiveEnergyQ2PhaseA;
+
+    /**
+     * Quadrant 2: Phase B imported reactive energy
+     */
+    public Optional<Long> acImportedReactiveEnergyQ2PhaseB;
+
+    /**
+     * Quadrant 2: Phase C imported reactive energy
+     */
+    public Optional<Long> acImportedReactiveEnergyQ2PhaseC;
+
+    /**
+     * Quadrant 3: Total exported reactive energy
+     */
+    public Optional<Long> acExportedReactiveEnergyQ3Total;
+
+    /**
+     * Quadrant 3: Phase A exported reactive energy
+     */
+    public Optional<Long> acExportedReactiveEnergyQ3PhaseA;
+
+    /**
+     * Quadrant 3: Phase B exported reactive energy
+     */
+    public Optional<Long> acExportedReactiveEnergyQ3PhaseB;
+
+    /**
+     * Quadrant 3: Phase C exported reactive energy
+     */
+    public Optional<Long> acExportedReactiveEnergyQ3PhaseC;
+
+    /**
+     * Quadrant 4: Total exported reactive energy
+     */
+    public Optional<Long> acExportedReactiveEnergyQ4Total;
+
+    /**
+     * Quadrant 4: Phase A exported reactive energy
+     */
+    public Optional<Long> acExportedReactiveEnergyQ4PhaseA;
+
+    /**
+     * Quadrant 4: Phase B exported reactive energy
+     */
+    public Optional<Long> acExportedReactiveEnergyQ4PhaseB;
+
+    /**
+     * Quadrant 4: Phase C exported reactive energy
+     */
+    public Optional<Long> acExportedReactiveEnergyQ4PhaseC;
```

Comment:
You could refactor a lot of code if you grouped all of the phase fields into a nested class. 

```
public MeterModelBlock{
...
    public PhaseBlock phaseA = new PhaseBlock();
    public PhaseBlock phaseB = new PhaseBlock();
    public PhaseBlock phaseC = new PhaseBlock();
...

    public static PhaseBlock {
...
        public Optional<Long> acExportedReactiveEnergyQ4;
...
    }
}
```
Then you could access all the data like so: `meterBlock.phaseC.acExportedReactiveEnergyQ4`.

Model parsing wouldn't be all that different, but where you would safe a lot of code is in this class and in the handler where you would only need a single method to update all the channels for a phase group, rather than having a method for each phase like you do now. 
WDYT?

Reply:
I was also hesitating a lot about that part, because of the lot of similar code. On the other hand I didn't want to overcomplicate the data structures, neither to push too much of knowledge into the parser class.
However you solution seems now a good compromise, so I will definitely give it a try.

Reply:
Thank you again for the idea, I could remove lots of lines. This happens when I'm too long into one solution :)

Reply:
Np, I just saw a lot of duplicate code and proposed a solution. :)